###the input data of SMR was generated by matrixeqtl package
###smr need chromomose is a numeric data
#Function -------------------------------------------------------------------------
smr_fun=function(exposure,outcome,out,threshold,gtf){
  inter="inter_result/"
  smr="../../home/software/smr-1.3.1-linux-x86_64/smr-1.3.1"
  #exposure="../input_V2/SMR/qtls.txt"
  #outcome="../input_V2/SMR/FT16mlm.mlma"


  ##load mlm result of output and delect some colums
  message("Converting mlm file to SMR format")
  suppressMessages(library(data.table))
  suppressMessages(library(stringr))
  print("make file")
  data=data.frame(fread(outcome))
  gwa=data
  data=data[,c(2,4,5,6,7,8,9,10)]
  names(data)=c("SNP","A1","A2","freq","b","se","p","n")
  write.table(data,paste0(inter,"smr_output.ma"),row.names = F,col.names = T,quote = F)

  ##make esi file
  message("Converting Omic QTL file to SMR format")
  system(paste("/usr/bin/python3 code/gff_format.py",gtf,paste0(inter,"smrgff.gff3")))
  gtf=data.frame(fread(paste0(inter,"smrgff.gff3"),fill=T))
  gene=data.frame(str_split_fixed(str_split_fixed(gtf$V9[which(gtf$V3=="gene")],";",2)[,1],":",2)[,2])
  gene$chr=gtf$V1[which(gtf$V3=="gene")]
  gene$BP=(gtf$V4[which(gtf$V3=="gene")]+gtf$V5[which(gtf$V3=="gene")])/2
  gene$start=gtf$V4[which(gtf$V3=="gene")]
  gene$end=gtf$V5[which(gtf$V3=="gene")]
  names(gene)[1]="Probe"
  row.names(gene)=gene$Probe
  gene=gene[grep("[0-9]+",gene[,2]),]
  
  qtls=data.frame(fread(exposure))
  qtls=qtls[qtls[,1]%in%gwa[,2],]
  qtls=qtls[qtls[,2]%in%gene[,1],]
  exposure=paste0(inter,"qtls_V2.txt")
  write.table(qtls,file=exposure,row.names = F,col.names =T,quote=F)  
  
  row.names(gwa)=gwa$SNP
  esi=data.frame(gwa[qtls$SNP,"Chr"],qtls$SNP,0,gwa[qtls$SNP,"bp"],gwa[qtls$SNP,"A1"],gwa[qtls$SNP,"A2"],gwa[qtls$SNP,"Freq"])


  ##make epi file
  print("make epi file")

  
  epi=data.frame(gene[qtls$gene,"chr"],qtls$gene,0,gene[qtls$gene,"BP"],qtls$gene,"+")
  nanum=is.na(esi[,4])|is.na(epi[,4])
  esi=esi[!nanum,]
  epi=epi[!nanum,]
  write.table(esi,paste0(inter,"new.esi"),row.names = F,col.names = F,quote = F)
  write.table(epi,paste0(inter,"new.epi"),row.names = F,col.names = F,quote = F)
  ##make gene list file
  glist=data.frame(gene$chr,gene$start,gene$end,gene$Probe)
  write.table(glist,paste0(inter,"glist.txt"),row.names = F,col.names = F,quote = F)
  ##make a BESD file from matrixeqtl file
  cmd_besd=paste(smr,"--eqtl-summary",exposure,"--matrix-eqtl-format","--make-besd","--out",
                 paste0(inter,"BESD"))
  write.table(cmd_besd,file="info.txt",append=T,row.names = F,col.names = F,quote=F)
  system(cmd_besd)

  
  ##update of BESD
  
  cmd_updata=paste(smr,"--beqtl-summary",paste0(inter,"BESD"),"--update-esi", paste0(inter,"new.esi"))
  write.table(cmd_updata,file="info.txt",append=T,row.names = F,col.names = F,quote=F)
  system(cmd_updata)
  cmd_updata=paste(smr,"--beqtl-summary",paste0(inter,"BESD"),"--update-epi", paste0(inter,"new.epi"))
  write.table(cmd_updata,file="info.txt",append=T,row.names = F,col.names = F,quote=F)
  system(cmd_updata)

  esi=data.frame(fread(paste0(inter,"BESD.esi")))
  nanum2=is.na(esi[,1])
  esi=esi[!nanum2,]
  write.table(esi,file=paste0(inter,"BESD.esi"),row.names = F,col.names = F,quote=F)

  if(threshold=="Bonferroni"){
    #threshold can calculated
    threshold=0.05/nrow(gwa)
  }else{
    threshold=as.numeric(threshold)
  }
  ##SMR
  
  message("Performing SMR")
  cmd_smr=paste(smr,"--bfile",paste0(inter,"bfile"),"--gwas-summary", paste0(inter,"smr_output.ma"),"--beqtl-summary",paste0(inter,"BESD"),"--peqtl-smr",threshold,"--thread-num 10",
                "--out",paste0(out,"smr"))
  write.table(cmd_smr,file="info.txt",append=T,row.names = F,col.names = F,quote=F)
  system(cmd_smr)
  message("The SMR result was saved in smr.smr")
  #generate plot file
  data=data.frame(fread(paste0(out,"smr",".smr")))
  write.table(sum(data$p_GWAS<threshold),file="info.txt",append=T,row.names = F,col.names = F,quote=F)
  if(sum(data$p_GWAS<threshold)>1){
    data=data[data$p_GWAS<threshold,]
    genename=data$Gene
    source("code/plot_SMR.r")
    write.table(genename,file="info.txt",append=T,row.names = F,col.names = F,quote=F)
    if(length(genename)>0){
      for(i in genename){
        cmd_smrplot=paste(smr,"--bfile",paste0(inter,"bfile"),"--gwas-summary", paste0(inter,"smr_output.ma"),"--beqtl-summary",paste0(inter,"BESD"),"--peqtl-smr",threshold,"--thread-num 10",
                          "--out",paste0(out,i,"_smr"),"--plot","--probe",i,"--probe-wind 500","--gene-list",paste0(inter,"glist.txt"))
        system(cmd_smrplot)
      }
      for(i in genename){
        message(paste0("Plotting SMR result of ",i))
        SMRData = ReadSMRData(paste0(out,"plot/",i,"_smr.",i,".txt"))
        png(paste0(out,i,"smr.png"),width=1000,height=1000)
        par(mfrow=c(2,2),mar=c(10, 10, 10, 10))
        layout(matrix(c(1,1,2,3),ncol=2,byrow = T),heights=c(1,1))
        try(SMRLocusPlot(data=SMRData, smr_thresh=threshold, heidi_thresh=0.05, plotWindow=50, max_anno_probe=10),silent = T)
        try(SMREffectPlot(data=SMRData, trait_name=""),silent = T)
        dev.off()
      }
    }
  }else{
    message("No significant SMR result")
    message("Maybe the over setted threshold")
    message("Please Check the .log file")
  }
}
# Code --------------------------------                   ------------------------------------

source("code/All_basic_function.R")
args <- commandArgs(TRUE)
print(args)
exposure=args[1]
outcome=args[2]
vcf=args[3]
gtf=args[4]
threshold=args[5]
out=args[6]
data_convert_fun(vcf)
smr_fun(exposure,outcome,out,threshold,gtf)


